### DISCUSSION TITLE
delay-us just freezes.  How should I be using it?

### DISCUSSION THREAD

**Discussion by Andy2No** [tabemann/zeptoforth#167] [Q&A] [answered]
Trying to familiarise myself with Zeptoforth, I tried writing simple loops with a delay in them, like in the Arduino blink sketch.  The closest I could find to delay(n) was delay-us, so I tried that, but whatever I do it just freezes until I type Control-C or press the rest button.

I'm using the current release, non-USB version on a Pi Pico RP2040, at the moment, installed with zeptoforth_full-1.9.1.uf2

The following freezes it every time.  I started by verifying that the microsecond timer is running;

```
timer import  ok
us-counter-lsb  ok
. 20150615  ok
us-counter-lsb  ok
. 24748677  ok
1000 delay-us
```

**Comment by tabemann** [comment]
On Sat, Dec 14, 2024 at 12:51 AM Andy2No ***@***.***> wrote:

> Trying to familiarise myself with Zeptoforth, I tried writing simple loops
> with a delay in them, like in the Arduino blink sketch. The closest I could
> find to delay(n) was delay-us, so I tried that, but whatever I do it just
> freezes until I type Control-C or press the rest button.
>
> I'm using the current release, non-USB version on a Pi Pico RP2040, at the
> moment, installed with zeptoforth_full-1.9.1.uf2
>
> The following freezes it every time. I started by verifying that the
> microsecond timer is running;
>
> timer import  ok
> us-counter-lsb  ok
> . 20150615  ok
> us-counter-lsb  ok
> . 24748677  ok
> 1000 delay-us
>
> DELAY-US takes a double-cell number of microseconds, so in this case you
would want:

1000. DELAY-US

rather than:

1000 DELAY-US

as the latter will pull two cells from the stack, using 1000 as the higher
cell and garbage beneath it as the lower cell, and thus wait too long and
corrupt the stack.

Travis

**Comment by tabemann** [comment, marked as answer]
On Sat, Dec 14, 2024 at 11:58 AM Travis Bemann ***@***.***> wrote:

> On Sat, Dec 14, 2024 at 12:51 AM Andy2No ***@***.***> wrote:
>
>> Trying to familiarise myself with Zeptoforth, I tried writing simple
>> loops with a delay in them, like in the Arduino blink sketch. The closest I
>> could find to delay(n) was delay-us, so I tried that, but whatever I do it
>> just freezes until I type Control-C or press the rest button.
>>
>> I'm using the current release, non-USB version on a Pi Pico RP2040, at
>> the moment, installed with zeptoforth_full-1.9.1.uf2
>>
>> The following freezes it every time. I started by verifying that the
>> microsecond timer is running;
>>
>> timer import  ok
>> us-counter-lsb  ok
>> . 20150615  ok
>> us-counter-lsb  ok
>> . 24748677  ok
>> 1000 delay-us
>>
>> DELAY-US takes a double-cell number of microseconds, so in this case you
> would want:
>
> 1000. DELAY-US
>
> rather than:
>
> 1000 DELAY-US
>
> as the latter will pull two cells from the stack, using 1000 as the higher
> cell and garbage beneath it as the lower cell, and thus wait too long and
> corrupt the stack.
>

Note that this is a spinwait, so it is good for short,  high-resolution
waits but not so good for longer waits. For longer waits you want either
the standardish Forth MS, which takes a single-cell number of milliseconds,
as in:

1 MS

or you can delay a specific task a particular number of ticks (100
microsecond increments) with DELAY ( delay start task -- ), which is in the
TASK module. This will delay the specified task (e.g. CURRENT-TASK in the
TASK module) to 'delay' ticks after the tick 'start' relative to
SYSTICK-COUNTER (in the SYSTICK module). Note that if the target task is on
a different core than the current task, the task may not be delayed
immediately.

As a use example consider:

TASK IMPORT
SYSTICK IMPORT
10 SYSTICK-COUNTER CURRENT-TASK DELAY

This will delay the current task to approximately 1 millisecond (10 ticks)
after the current tick.

Note that MS and DELAY are multitasker-friendly, as they will yield control
to other tasks so they can execute rather than spinning. However, they are
not as high-resolution as DELAY-US .

Travis

**Comment by Andy2No** [comment]
Thanks.  I was initially looking for a delay(milliseconds) equivalent, but delay-us was all I could find.

As a first time user of the documentation in zeptoforth.epub, none of that was clear to me.  I see it says

```
delay-us
( us-d – )
```
but I can't find anything that explains that notation to me.  It seems to me that using double word parameters is probably rare enough to spell that out, in so many words.

I temporarily settled on doing this instead:

```
: zz 10000 * 0 do loop ; ( Short delay, CPU/MHz dependent )
```
which is roughly in the ball park of an n millisecond wait, on a Pi Pico.

**Comment by tabemann** [comment]
On Sat, Dec 14, 2024 at 12:53 PM Andy2No ***@***.***> wrote:

> Thanks. I was initially looking for a delay(milliseconds) equivalent, but
> delay-us was all I could find.
>
> As a first time user of the documentation in zeptoforth.epub, none of that
> was clear to me. I see it says
>
> delay-us
> ( us-d – )
>
> but I can't find anything that explains that notation to me. It seems to
> me that using double word parameters is probably rare enough to spell that
> out, in so many words.
>
> I temporarily settled on doing this instead:
>
> : zz 10000 * 0 do loop ; ( Short delay, CPU/MHz dependent )
>
> which is roughly in the ball park of an n millisecond wait, on a Pi Pico.
>

Forgive my notation - 'us-d' stands for microseconds ('us' for short),
double-cell (which in Forth is commonly abbrevated as 'd'; e.g. in local
definitions double-cell locals are precded with 'D:').

Travis

**Reply by Andy2No** [reply]
Thanks for the clarification.
